ambry-cloud/src/main/java/com/github/ambry/cloud/HelixVcrClusterParticipant.java

```diff
@@ -105,21 +117,44 @@ public HelixVcrClusterParticipant(CloudConfig cloudConfig, ClusterMapConfig clus
    * @param partitionIdStr The partitionIdStr notified by Helix.
    */
   public void addPartition(String partitionIdStr) {
-    PartitionId partitionId = partitionIdMap.get(partitionIdStr);
+    // Add partition with retry logic in case Ambry clustermap is not update to date.
+    retryExecutor.runWithRetries(retryPolicy, callback -> doAddPartition(partitionIdStr, callback), e -> true,
+        (result, exception) -> {
+          if (exception != null) {
+            logger.warn("Partition {} is not in clusterMap on addPartition and retry failed.", partitionIdStr,
+                exception);
+          } else {
+            logger.info("Partition {} is added to current VCR: {}. Number of assigned partitions: {}", partitionIdStr,
+                vcrInstanceName, assignedPartitionIds.size());
+          }
+        });
+  }
+
+  private void doAddPartition(String partitionIdStr, Callback<Object> callback) {
+    byte[] partitionBytes =
+        ClusterMapUtils.serializeShortAndLong(AmbryPartition.CURRENT_VERSION, Long.parseLong(partitionIdStr));
+    PartitionId partitionId = null;
+    try {
+      partitionId = clusterMap.getPartitionIdFromStream(new ByteBufferInputStream(ByteBuffer.wrap(partitionBytes)));
+    } catch (IOException e) {
+      callback.onCompletion(null, e);
+      return;
+    }
     if (partitionId != null) {
       if (assignedPartitionIds.add(partitionId)) {
         for (VcrClusterParticipantListener listener : listeners) {
```

Comment:
We are adding the new partition id to the `assignedPartitionIds `, and only when not already present, we call `onPartitionAdded `. Since we are retrying, this can cause issues. In the first try the `assignedPartitionIds.add` will succeed and `onPartitionAdded ` might fail. In the subsequent retries `assignedPartitionIds.add` will already have the partition id, and it will return false. This will basically make the retries no-op.

Reply:
good catch. I am trying to get rid of `assignedPartitionIds `. Will update the logic here.